druid/src/widget/textbox/editable_text.rs

```diff
@@ -0,0 +1,212 @@
+// Copyright 2019 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Traits for text editing and a basic String implementation.
+
+use crate::unicode_segmentation::GraphemeCursor;
+
+pub trait EditableText {
+    // pub fn edit<T: IntervalBounds>(&mut self, iv: T, new: String);
+
+    // pub fn slice<T IV>(&self, iv: T) -> String;
+
+    // pub fn is_codepoint_boundary(&self, offset: usize) -> bool;
+
+    // fn prev_codepoint_offset(&self, offset: usize) -> Option<usize>;
+
+    // fn next_codepoint_offset(&self, offset: usize) -> Option<usize>;
+
+    // pub fn at_or_next_codepoint_boundary(&self, offset: usize) -> Option<usize>;
+
+    // pub fn at_or_prev_codepoint_boundary(&self, offset: usize) -> Option<usize>;
+
+    fn prev_grapheme_offset(&self, offset: usize) -> Option<usize>;
+
+    fn next_grapheme_offset(&self, offset: usize) -> Option<usize>;
+}
+
+impl EditableText for &str {
```

Comment:
thinking about this more, I don't know if makes sense for us to think of a string slice as 'editable', since it generally isn't. (For instance, you can't change the size of a string slice).

In fact, I think we can get away with having this only impl'd for `String`. Because the trait defines functions as taking `&self` and `&mut self`, it doesn't make much sense to impl it for `&String`; in that case the `self` type in the methods would be `&&String` and `&mut &String`.